### DISCUSSION TITLE
Reusing image assets to reduce file size

### DISCUSSION THREAD

**Discussion by 5rEUKk5Xa4yh** [QuestPDF/QuestPDF#386] [Q&A] [answered]
Hi,
I'm trying to migrate to using QuestPDF. My use-case is farily simple, so I'm sure it's possible in the library but cannot figure out how to achieve this. I want to utilise resource sharing, xobjects in pdf, to produce PDF files with many pages but small file sizes.

In my output, i will be only drawing images to the document's pages. Each page will have at least 4 images. Many images will be unique to a page (a barcode, for example) but many will be shared and shared randomly. 

A simple example: 
I have 2 images: image1.png and image2.png. Both are 1MB in size.

I create 10 pages and need to draw images in this order:
P1: image1.png
P2: image2.png
P3 image1.png
P4: image1.png
P5: image1.png
P6: image2.png
P7: image1.png
P8: image2.png
P9: image2.png
P10: image2.png

How can I produce a 10-page document that is ~2MB in size instead of ~10MB?

I have tried feeding bytes and streams to the image method. I have tried instantiating those objects outside the loop in case Quest can use object reference matching to determine duplicate content. I have tried utilising components but I always get a file that is numberOfImages x numberOfPages in size. 
I have manged to get the desired file size by using Row().ConstantItem(). But this is not really usable for me. 

I will get a list of images to place on a document and they will have coordinates associated with them. I cannot assume any page's layout will be the same as another.

An ideal workflow would be as below, and require no intervention from the user:

//Adding 10 pages
for (var i = 0;  i < 10;  i++)
{
  container.Page(page =>
  {
  page.Content()
  .PaddingVertical(1, Unit.Centimetre)
     . Column(column =>
       {
           column.Item().TranslateX(50).TranslateY(-250).Image(imageStream); //If this stream has been used before, reuse asset
       }); 
  });
}

I have tried PdfPig and can achieve this by working around its image add method which returns a reference to an AddedImage object - this can then be added to subsequent pages and will use the xobject reference. Syncfusion PDF has its own image object and will detect collisions based on those references.

I've searched this board and watched the jetbrains presentation, but am at a loss as to how to implement this. I feel like it's too obvious so isn't noted down anywhere and i must be missing something.

**Comment by MarcinZiabek** [comment]
Hi 😁 Thank you for noticing this isssue and reporting it. 

I haven't got a change to investigate how exactly SkiaSharp handles this scenario. However, it is pretty possible that for each Column item, a brand new image asset is created. I am also not sure, how well SkiaSharp handles a single image that is drawn multiple times. I will need to perform some tests later.

Let's assume that everything goes well and this is truly QuestPDF problem. We have two choices:
1) Create a `SharedImage` class, that any developer needs to use in order to reuse the image, and therefore reduce the PDF size. This is quite simple to implement, but not discoverable not obvious.
2) Try to find a way to detect this type of scenario and perform it automatically. Detecting duplicate calls to the `Image` method with same fileName is possible, I am not sure about Streams though... Not to mention, that images may live in difference scopes. For some images we want to cache them in memory, some require shorter lifetimes.

Unfortunately, I don't have much spare time recently. Therefore, I am sorry for responding so late.

**Reply by MarcinZiabek** [reply]
I can confirm that SkiaSharp is capable of reusing the same SkImage.

Generally, I am against surfacing up the SkiaSharp dependency. Much rather, I would like to prepare a simple QuestPDF.Image class, or some kind of wrapper. Also, image caching may be a nice addition for performance benefits.

**Comment by wbijker** [comment]
Hi @5rEUKk5Xa4yh 

I had exact the same problem. Given the fact that SKImage is sharable between multiple pages, you can use a Dynamic component to render the different SkImages per page using canvas.

In your scenario, something like this:


```c#
    public class DynamicBackgroundImages : IDynamicComponent<int>
    {
        // Load images once into memory and keep the references.
        // SkiaSharp is clever enough to reuse the image with the same reference 
        private static readonly SKImage Image1 = SKImage.FromEncodedData("image1.png");
        private static readonly SKImage Image2 = SKImage.FromEncodedData("image2.png");
        
        // determine what image should render on what page. Any order or logic you want
        private static SKImage[] PageImage = new[]
        {
            Image1,
            Image2,
            Image1,
            Image1,
            Image2,
            Image1,
            Image2,
            Image2,
            Image2,
        };
        
        public DynamicComponentComposeResult Compose(DynamicContext context)
        {
            var content = context.CreateElement(container =>
            {
                container.Canvas((canvas, space) =>
                {
                    // stretch image 
                    var dest = SKRect.Create(0, 0, space.Width, space.Height);
                    canvas.DrawImage(PageImage[(context.PageNumber - 1) % PageImage.Length], dest);
                });
            });

            return new DynamicComponentComposeResult()
            {
                HasMoreContent = false,
                Content = content
            };
        }

        public int State { get; set; }
    }

  // the usage somewhere.
  page.Background().Dynamic(new DynamicBackgroundImages());

```
The result PDF will only embed the images once. I have tested with 100 pages and size is a slightly more than image1 and image2 combined.

**Comment by DawidRobertNowak** [comment, marked as answer]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Since this discussion was opened, QuestPDF now natively supports shared images — images loaded from file paths are automatically cached and embedded only once in the PDF, regardless of how many pages they appear on. You can find the documentation here:
https://www.questpdf.com/api-reference/image/shared.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
